feat!: Thread safe hooks, provider, and context

kinyoklion opened the pull request:
- Change storage type of global hooks to a concurrent stack.
    - Should be able to add and remove from this without locking,
- Protect references used from multiple threads with a Read/Write lock.
- Change read only collection types to immutable collection types.
    - Readonly collections are just views into existing mutable collections. So they are still subject to the same modification problems. Immutable collections are supposed to be fixed and unchanging. Considering they are an interface you could still violate this constraint, but it is less likely.

Related to: https://github.com/open-feature/dotnet-sdk/issues/56

kinyoklion commented on src/OpenFeatureSDK/OpenFeature.cs:
```diff
@@ -53,25 +61,30 @@ public sealed class OpenFeature
-        /// <returns>A immutable list of <see cref="Hook"/></returns>
-        public IReadOnlyList<Hook> GetHooks() => this._hooks.AsReadOnly();
+        /// <returns>Enumeration of <see cref="Hook"/></returns>
+        public IEnumerable<Hook> GetHooks() => this._hooks.Reverse();
```
In the open feature API and Client, the SDK controls the implementation type and can promise the use of a concurrent collection.

In the provider this is not the case, so that is constrained to an immutable collection.

kinyoklion replied:
Also, I am reversing these for now. I wasn't completely comfortable with all the test changes required for them to be in the opposite order.

kinyoklion commented on src/OpenFeatureSDK/FeatureProvider.cs:
```diff
@@ -22,8 +21,8 @@ public abstract class FeatureProvider
-        /// <returns></returns>
-        public virtual IReadOnlyList<Hook> GetProviderHooks() => Array.Empty<Hook>();
+        /// <returns>Immutable list of hooks</returns>
+        public virtual IImmutableList<Hook> GetProviderHooks() => ImmutableList<Hook>.Empty;
```
Alternatively this could be enumerable like other places, with documentation that the provider must use a thread-safe implementation. Technically a readonlylist, with the same constraint, would also works. I think immutable directs people in a way that the easiest solution is also the correct solution.

toddbaert replied:
> I think immutable directs people in a way that the easiest solution is also the correct solution.

Agreed.

kinyoklion commented on src/OpenFeatureSDK/OpenFeatureClient.cs:
```diff
@@ -225,9 +228,9 @@ public FeatureClient(FeatureProvider featureProvider, string name, string versio
+                .Concat(this.GetHooks())
                 .Concat(options?.Hooks ?? Enumerable.Empty<Hook>())
-                .Concat(this._featureProvider.GetProviderHooks())
+                .Concat(OpenFeature.Instance.GetProvider().GetProviderHooks())
```
Ok. This isn't correct. You would't want the hooks from a different provider than the evaluation method. Those two things need to be resolved at the same time.

kinyoklion replied:
I think that we need to get the reference, then use that reference to get the method and the hooks. We would not just want to get them in the call, because they could still be separated in time.

kinyoklion commented on src/OpenFeatureSDK/OpenFeatureClient.cs:
```diff
@@ -17,11 +18,32 @@ namespace OpenFeatureSDK
         private readonly ILogger _logger;
         private EvaluationContext _evaluationContext;
 
+        /// <summary>
```
This is a little cumbersome. I am not sure if it is more or less cumbersome than the alternative.
The alternative would be to convert all the methods from expression bodies to normal function bodies, then alias the provider and method, then pass those to evaluate.

kinyoklion commented on src/OpenFeatureSDK/OpenFeature.cs:
```diff
@@ -30,19 +36,41 @@ public sealed class OpenFeature
+        public void SetProvider(FeatureProvider featureProvider)
+        {
+            this._featureProviderLock.EnterWriteLock();
+            try
```
None of these code blocks can throw, but when using this type of lock I think using try/finally is a customary way to release the lock.

toddbaert replied:
Yep, agreed. In fact, I made a little lock-wrapper [class](https://github.com/open-feature/java-sdk/pull/131/files#diff-f7f996e84ae9d77890aa5aa0cde5a7e3b08401b43173268bc134083536ea26dc) that allows us to use a try-with-resources block in Java (I think c# has an equivalent with the `using` keyword and `IDisposable` implementations).

kinyoklion replied:
Especially for double locks, where they could nest the using statement.

kinyoklion commented on src/OpenFeatureSDK/OpenFeature.cs:
```diff
@@ -30,19 +36,53 @@ public sealed class OpenFeature
+        /// operations. For instance, during an evaluation the flag resolution method, and the provider hooks
+        /// should be accessed from the same reference, not two independent calls to
+        /// <see cref="OpenFeature.GetProvider"/>.
+        /// </para>
```
One thing I am not calling out here is that setting hooks, provider, and context are not guaranteed to be ordered among themselves. If you have one thread setting all 3, and another thread doing evaluations, then a single evaluation may use some combination of the values being set.

toddbaert replied:
> One thing I am not calling out here is that setting hooks, provider, and context are not guaranteed to be ordered among themselves. If you have one thread setting all 3, and another thread doing evaluations, then a single evaluation may use some combination of the values being set.

That makes sense, and I think mentioning this might be adding cognitive overload with not-too-much value returned.

As for the block comment, I had added something similar in the read-lock block in Java, so it makes sense to me and I think it's worth pointing out.

kinyoklion replied:
So, here each of the 3 things are individually thread safe. 

The provider will use a stable reference, which covers the provider hooks and other data.

The context is immutable, and we get a reference at a specific point in time. The reference globally may change, but we are decoupled from that because we got a reference. Locking for the duration of getting the context and the provider wouldn't prevent interleaved setting of them. You could set one, and then the evaluation gets the lock, you have to wait, and then set the other. 

The last bit are the other two sets of hooks. These are interesting. There isn't anything which inherently couples them to anything else. They are copied, so the list will be stable in the places they are iterate, but up until they are iterated, they could change.

If you have cross dependent hooks then they should be added with AddHooks, which will add them atomically. I should document that.

kinyoklion replied:
I added a couple comments to the hook methods.

toddbaert replied:
> If you have cross dependent hooks

Worth saying I generally think this should be avoided, but I think people might do it.

toddbaert replied:
> Locking for the duration of getting the context and the provider wouldn't prevent interleaved setting of them.

Wouldn't read-locking in the evaluation on a shared lock which is write-locked by both the provider and context mutators address exactly this? Or am I misunderstanding?

kinyoklion replied:
Not unless you also locked for the duration of setting both of them.

If they can be independently set, as they can in the implementation at the moment, then it doesn't.

```
Thread 1:
SetProvider(provider)
  writeLock(_provider)
// Evaluation could start here
SetContext(context)
  writeLock(_context)

Thread 2:
Evaluate()
  readLock(_provider)
    readLock(_context)
```

So, given this code someone could intend to set the provider and context, but it would be possible for evaluation to grab the lock before thread 1 set the context, but after it set the provider.

In order to not interleave it would need to be more like

```
Thread 1:
writeLock(_provider)
  writeLock(_context)
    setProvider
    setContext
    
Thread 2:
Evaluate()
  readLock(_provider)
    readLock(_context)
```

Or use a single lock

toddbaert replied:
Ahh, yes, I agree. I misunderstood your initial statement.

kinyoklion commented on src/OpenFeatureSDK/OpenFeatureClient.cs:
```diff
@@ -241,7 +273,7 @@ public FeatureClient(FeatureProvider featureProvider, string name, string versio
                 defaultValue,
                 flagValueType, this._metadata,
-                OpenFeature.Instance.GetProviderMetadata(),
+                provider.GetMetadata(),
```
We don't want to call `GetProviderMetadata` here on the `OpenFeature` singleton, because it could have change between when we got the resolution method and here. We want to be sure to only use the cached reference.

kinyoklion commented on src/OpenFeatureSDK/OpenFeatureClient.cs:
```diff
@@ -36,15 +58,13 @@ public sealed class FeatureClient : IFeatureClient
         /// <param name="logger">Logger used by client</param>
         /// <param name="context">Context given to this client</param>
         /// <exception cref="ArgumentNullException">Throws if any of the required parameters are null</exception>
-        public FeatureClient(FeatureProvider featureProvider, string name, string version, ILogger logger = null, EvaluationContext context = null)
```
Just realized I need to re-introduce the equivalent of this check elsewhere.

toddbaert commented on src/OpenFeatureSDK/OpenFeatureClient.cs:
```diff
@@ -17,11 +18,38 @@ namespace OpenFeatureSDK
+            if (provider == null)
+            {
+                provider = new NoOpFeatureProvider();
+                this._logger.LogDebug("No provider configured, using no-op provider");
```
I think No-op is a valid configuration, and may be used intentionally, especially if OpenFeature is implemented in 3rd party libraries.

For this reason, I think this should remain a `debug` level log.

toddbaert commented on src/OpenFeatureSDK/OpenFeature.cs:
```diff
@@ -30,19 +36,53 @@ public sealed class OpenFeature
         /// </summary>
         /// <returns><see cref="FeatureProvider"/></returns>
-        public FeatureProvider GetProvider() => this._featureProvider;
+        public FeatureProvider GetProvider()
+        {
+            this._featureProviderLock.EnterReadLock();
+            try
+            {
+                return this._featureProvider;
+            }
+            finally
+            {
+                this._featureProviderLock.ExitReadLock();
+            }
+        }
```
Is it necessary to lock this? I think as long as we lock during it's retrieval in the evaluation, that's sufficient, right?

kinyoklion replied:
I think there is another consideration here, in that the getters are public, so they maybe should still have ordering guarantees. 

I think there is a way to allow re-entrant read locks though, so if we did need to lock for multiple operations, then we could change to the re-entrant form. Allowing acquiring read locks inside read locks.

kinyoklion replied:
https://learn.microsoft.com/en-us/dotnet/api/system.threading.readerwriterlockslim?view=net-6.0#entering-the-lock-recursively

toddbaert commented on src/OpenFeatureSDK/OpenFeature.cs:
```diff
@@ -81,13 +127,41 @@ public sealed class OpenFeature
-        /// <returns></returns>
-        public EvaluationContext GetContext() => this._evaluationContext;
+        /// <returns>An <see cref="EvaluationContext"/></returns>
+        public EvaluationContext GetContext()
+        {
+            this._evaluationContextLock.EnterReadLock();
+            try
+            {
+                return this._evaluationContext;
+            }
+            finally
+            {
+                this._evaluationContextLock.ExitReadLock();
+            }
+        }
```
Same as: https://github.com/open-feature/dotnet-sdk/pull/79/files#r992477160

toddbaert commented on src/OpenFeatureSDK/OpenFeatureClient.cs:
```diff
@@ -17,11 +18,38 @@ namespace OpenFeatureSDK
+        private (Func<string, T, EvaluationContext, Task<ResolutionDetails<T>>>, FeatureProvider)
+            ExtractProvider<T>(
+                Func<FeatureProvider, Func<string, T, EvaluationContext, Task<ResolutionDetails<T>>>> method)
+        {
```
It may be worth read-locking during this operation, and include the hooks and context here. The way this is implemented right now, evaluation context and hooks can change during the process of evaluation, though you've handled the provider.

kinyoklion commented:
I realized that I missed the client level context locking. So I have added that.

kinyoklion commented:
I've noticed another thing that needs to be resolved. The ReaderWriterLock type is disposable, so the containing type should implement disposable. Not really a problem for the global OpenFeature instance, because it isn't going anywhere. It is inconvenient for the client instances to be disposable though.

So I may want to consider a normal lock in that case.

benjiro reviewed (approved):
LGTM 👍🏻

kinyoklion commented:
@toddbaert @benjiro I switched the OpenFeatureClient locks to normal locks. This does mean if you are using one client from many threads, that it will have more contention. Maybe... make more clients.

The alternative would be to make it disposable, but that then becomes like a, "what color are your functions" kind of situation.

In OpenFeature.cs I am not disposing. I added a comment. The singleton isn't meant to go away, so disposing it doesn't make sense.

toddbaert commented:
I know there's been a few changes on this, but I think it's this is good work and certainly more threadsafe than what we had before. I'm in favor of merging it unless there's objections.
